Frame rate (through FrameDurationLimits) does not work at node startup

In the `readme.md` it is indicated that the framerate can be controlled using `FrameDurationLimits` dynamic parameter.

Running this:
```
ros2 run camera_ros camera_node --ros-args -p camera:=0 -p format:="RGB888" -p width:=640 -p height:=480 -p FrameDurationLimits:="[20000,20000]"
```
does not change the frame rate. Checking the topic frequency results in:
```
(ros2) mh5@mh5g:~ $ ros2 topic hz /camera/image_raw
...
average rate: 17.695
	min: 0.054s max: 0.057s std dev: 0.00065s window: 19
average rate: 17.665
	min: 0.054s max: 0.057s std dev: 0.00049s window: 37
average rate: 17.659
	min: 0.054s max: 0.057s std dev: 0.00041s window: 55
average rate: 17.656
	min: 0.054s max: 0.057s std dev: 0.00036s window: 73
average rate: 17.653
	min: 0.054s max: 0.057s std dev: 0.00032s window: 91
average rate: 17.652
	min: 0.054s max: 0.057s std dev: 0.00030s window: 109
```
Explicitly changing the parameter with:
```
(ros2) mh5@mh5g:~ $ ros2 param set camera FrameDurationLimits "[20000,20000]"
Set parameter successful
```
results in:
```
...
average rate: 46.054
	min: 0.017s max: 0.039s std dev: 0.00677s window: 189
average rate: 47.451
	min: 0.017s max: 0.039s std dev: 0.00610s window: 243
average rate: 48.389
	min: 0.017s max: 0.039s std dev: 0.00559s window: 297
average rate: 49.061
	min: 0.017s max: 0.039s std dev: 0.00519s window: 351
average rate: 49.565
	min: 0.017s max: 0.039s std dev: 0.00486s window: 405
average rate: 49.958
	min: 0.017s max: 0.039s std dev: 0.00459s window: 459
average rate: 50.174
	min: 0.017s max: 0.039s std dev: 0.00444s window: 512
average rate: 50.441
	min: 0.017s max: 0.039s std dev: 0.00424s window: 566
```
I think it would be better if there is a "classic" node parameter called `framerate` that is used to setup the camera frame duration at node startup.

@christianrauch , Yes, that fixed the error and now seems to be fine:

```
(ros2) mh5@mh5g:~/mh5_ws $ ros2 run camera_ros camera_node --ros-args -p camera:=0 -p format:="RGB888" -p width:=640 -p height:=480 -p FrameDurationLimits:="[20000,20000]"
[0:25:15.059030448] [4207]  INFO Camera camera_manager.cpp:325 libcamera v0.3.2+99-1230f78d
[0:25:15.074021480] [4222]  INFO RPI pisp.cpp:695 libpisp version v1.0.7 28196ed6edcf 29-08-2024 (16:33:32)
[0:25:15.084034464] [4222]  INFO RPI pisp.cpp:1154 Registered camera /base/axi/pcie@120000/rp1/i2c@88000/ov5647@36 to CFE device /dev/media0 and ISP device /dev/media1 using PiSP variant BCM2712_C0
[0:25:15.084258353] [4222]  INFO RPI pisp.cpp:695 libpisp version v1.0.7 28196ed6edcf 29-08-2024 (16:33:32)
[0:25:15.093381558] [4222]  INFO RPI pisp.cpp:1154 Registered camera /base/axi/pcie@120000/rp1/i2c@80000/ov5647@36 to CFE device /dev/media4 and ISP device /dev/media2 using PiSP variant BCM2712_C0
[WARN] [1733674539.034751360] [camera]: stream configuration adjusted from "640x480-RGB888" to "640x480-RGB888"
[0:25:15.094171300] [4207]  INFO Camera camera.cpp:1197 configuring streams: (0) 640x480-RGB888
[0:25:15.094322004] [4222]  INFO RPI pisp.cpp:1450 Sensor: /base/axi/pcie@120000/rp1/i2c@88000/ov5647@36 - Selected sensor format: 640x480-SGRBG10_1X10 - Selected CFE format: 640x480-PC1G
[INFO] [1733674539.035350991] [camera]: camera "/base/axi/pcie@120000/rp1/i2c@88000/ov5647@36" configured with 640x480-RGB888 stream
[WARN] [1733674539.035663862] [camera]: control CnnEnableInputTensor (20007) not handled
[WARN] [1733674539.035793177] [camera]: control NoiseReductionMode (10002) not handled
[INFO] [1733674539.343784124] [camera]: using default calibration URL
[INFO] [1733674539.343847513] [camera]: camera calibration URL: file:///home/mh5/.ros/camera_info/ov5647__base_axi_pcie_120000_rp1_i2c_88000_ov5647_36_640x480.yaml
[ERROR] [1733674539.343906643] [camera_calibration_parsers]: Unable to open camera calibration file [/home/mh5/.ros/camera_info/ov5647__base_axi_pcie_120000_rp1_i2c_88000_ov5647_36_640x480.yaml]
[WARN] [1733674539.343915384] [camera]: Camera calibration file /home/mh5/.ros/camera_info/ov5647__base_axi_pcie_120000_rp1_i2c_88000_ov5647_36_640x480.yaml not found
```
And the ros topic:
```
(ros2) mh5@mh5g:~ $ ros2 topic hz camera/image_raw
average rate: 37.651
	min: 0.017s max: 0.072s std dev: 0.01153s window: 41
average rate: 42.266
	min: 0.017s max: 0.072s std dev: 0.00947s window: 89
average rate: 45.798
	min: 0.017s max: 0.072s std dev: 0.00783s window: 143
average rate: 47.591
	min: 0.017s max: 0.072s std dev: 0.00681s window: 197
average rate: 48.680
	min: 0.017s max: 0.072s std dev: 0.00610s window: 251
average rate: 49.411
	min: 0.017s max: 0.072s std dev: 0.00558s window: 305
average rate: 49.798
	min: 0.017s max: 0.072s std dev: 0.00526s window: 358
average rate: 46.429
	min: 0.017s max: 0.076s std dev: 0.00939s window: 381
```

While I understand the relation between the FrameDurationLimits and the frame rate, I am a little concerned about the fact that the publishing of the image topics is done not due to a ROS schedule, but due to a processing within the camera and libcamera. You can see that it seems to vary quite a lot and also seems to be off the expected 50Hz. Maybe we ca dissociate the two things:the FrameDurationLimits to control the exposure time and a separate parameter, used to control the publish rate of the the image messages, following the example of most ROS nodes. What do you think? I can setup another issue to cover this, so that this one can be left closed.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -42,20 +42,23 @@ set_property(TARGET utils PROPERTY POSITION_INDEPENDENT_CODE ON)
 # composable ROS2 node
 add_library(camera_component SHARED src/CameraNode.cpp)
 rclcpp_components_register_node(camera_component PLUGIN "camera::CameraNode" EXECUTABLE "camera_node")
-
-ament_target_dependencies(
-  camera_component
+set(PACKAGE_DEPENDENCIES
   "rclcpp"
   "rclcpp_components"
   "sensor_msgs"
   "camera_info_manager"
   "cv_bridge"
 )
-
+ament_target_dependencies(camera_component PUBLIC ${PACKAGE_DEPENDENCIES})
 target_include_directories(camera_component PUBLIC ${libcamera_INCLUDE_DIRS})
-target_link_libraries(camera_component ${libcamera_LINK_LIBRARIES} utils)
+target_link_libraries(camera_component PUBLIC ${libcamera_LINK_LIBRARIES})
+target_link_libraries(camera_component PRIVATE utils)
+
+ament_export_targets(camera_componentTargets HAS_LIBRARY_TARGET)
+ament_export_dependencies(${PACKAGE_DEPENDENCIES})
 
 install(TARGETS camera_component
+  EXPORT camera_componentTargets
   DESTINATION lib)
 
 install(DIRECTORY

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -43,9 +43,6 @@ set_property(TARGET utils PROPERTY POSITION_INDEPENDENT_CODE ON)
 add_library(camera_component SHARED src/CameraNode.cpp)
 rclcpp_components_register_node(camera_component PLUGIN "camera::CameraNode" EXECUTABLE "camera_node")
 
-target_include_directories(camera_component PUBLIC
-  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
-  $<INSTALL_INTERFACE:include>)
 ament_target_dependencies(
   camera_component
   "rclcpp"

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -115,7 +115,8 @@ private:
   std::unordered_map<unsigned int, libcamera::ControlValue> parameters;
   // keep track of set parameters
   ParameterMap parameters_full;
-  std::mutex parameters_lock;
+  std::mutex parameters_mutex;
+  std::unique_lock<std::mutex> parameters_lock {parameters_mutex, std::defer_lock};
   // compression quality parameter
   std::atomic_uint8_t jpeg_quality;
 

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -19,7 +19,6 @@
 #include <cv_bridge/cv_bridge.h>
 #endif
 #include <atomic>
-#include <functional>
 #include <iostream>
 #include <libcamera/base/shared_fd.h>
 #include <libcamera/base/signal.h>
@@ -62,7 +61,6 @@
 #include <string>
 #include <sys/mman.h>
 #include <thread>
-#include <tuple>
 #include <unordered_map>
 #include <utility>
 #include <vector>

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -571,8 +571,8 @@ CameraNode::declareParameters()
     }
     else {
       declare_parameter(id->name(), value, param_descr);
-      parameters_init[id->name()] = value;
     }
+    parameters_init[id->name()] = value;
   }
 
   // register callback to handle parameter changes

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -87,7 +87,8 @@ private:
   std::shared_ptr<libcamera::FrameBufferAllocator> allocator;
   std::vector<std::unique_ptr<libcamera::Request>> requests;
   std::vector<std::thread> request_threads;
-  std::unordered_map<libcamera::Request *, std::unique_ptr<std::mutex>> request_locks;
+  std::unordered_map<const libcamera::Request *, std::mutex> request_mutexes;
+  std::unordered_map<const libcamera::Request *, std::condition_variable> request_condvars;
   std::atomic<bool> running;
 
   struct buffer_info_t
@@ -424,8 +425,9 @@ CameraNode::CameraNode(const rclcpp::NodeOptions &options) : Node("camera", opti
   // create a processing thread per request
   running = true;
   for (const std::unique_ptr<libcamera::Request> &request : requests) {
-    request_locks[request.get()] = std::make_unique<std::mutex>();
-    request_locks[request.get()]->lock();
+    // create mutexes in-place
+    request_mutexes[request.get()];
+    request_condvars[request.get()];
     request_threads.emplace_back(&CameraNode::process, this, request.get());
   }
 
@@ -446,8 +448,18 @@ CameraNode::CameraNode(const rclcpp::NodeOptions &options) : Node("camera", opti
 
 CameraNode::~CameraNode()
 {
+  // stop request callbacks
+  for (std::unique_ptr<libcamera::Request> &request : requests)
+    camera->requestCompleted.disconnect(request.get());
+
   // stop request processing threads
   running = false;
+
+  // unlock all threads
+  for (auto &[req, condvar] : request_condvars)
+    condvar.notify_all();
+
+  // wait for all currently running threads to finish
   for (std::thread &thread : request_threads)
     thread.join();
 
@@ -588,15 +600,20 @@ CameraNode::declareParameters()
 void
 CameraNode::requestComplete(libcamera::Request *const request)
 {
-  request_locks[request]->unlock();
+  std::unique_lock lk(request_mutexes.at(request));
+  request_condvars.at(request).notify_one();
 }
 
 void
 CameraNode::process(libcamera::Request *const request)
 {
-  while (running) {
+  while (true) {
     // block until request is available
-    request_locks[request]->lock();
+    std::unique_lock lk(request_mutexes.at(request));
+    request_condvars.at(request).wait(lk);
+
+    if (!running)
+      return;
 
     if (request->status() == libcamera::Request::RequestComplete) {
       assert(request->buffers().size() == 1);

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -66,7 +66,7 @@
 #include <unordered_map>
 #include <utility>
 #include <vector>
-namespace enc = sensor_msgs::image_encodings;
+
 namespace rclcpp
 {
 class NodeOptions;
@@ -163,6 +163,8 @@ compressImageMsg(const sensor_msgs::msg::Image &source,
                  sensor_msgs::msg::CompressedImage &destination,
                  const std::vector<int> &params = std::vector<int>())
 {
+  namespace enc = sensor_msgs::image_encodings;
+
   std::shared_ptr<CameraNode> tracked_object;
   cv_bridge::CvImageConstPtr cv_ptr = cv_bridge::toCvShare(source, tracked_object);
 

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -590,8 +590,10 @@ CameraNode::declareParameters()
     RCLCPP_WARN_STREAM(get_logger(), s);
 
   std::vector<rclcpp::Parameter> parameters_init_list;
-  for (const auto &[name, value] : parameters_init)
-    parameters_init_list.emplace_back(name, value);
+  for (const auto &[name, value] : parameters_init) {
+    if (value.get_type() != rclcpp::ParameterType::PARAMETER_NOT_SET)
+      parameters_init_list.emplace_back(name, value);
+  }
   const rcl_interfaces::msg::SetParametersResult param_set_result =
     set_parameters_atomically(parameters_init_list);
   if (!param_set_result.successful)

--- a/src/parameter_conflict_check.cpp
+++ b/src/parameter_conflict_check.cpp
@@ -13,7 +13,9 @@ resolve_conflicts(const ParameterMap &parameters_default, const ParameterMap &pa
   // must not be enabled at the same time
 
   // default: prefer auto exposure
-  if (parameters_init.count("AeEnable") && parameters_init.at("AeEnable").get<bool>() &&
+  if (parameters_init.count("AeEnable") &&
+      (parameters_init.at("AeEnable").get_type() != rclcpp::ParameterType::PARAMETER_NOT_SET) &&
+      parameters_init.at("AeEnable").get<bool>() &&
       parameters_init.count("ExposureTime"))
   {
     // disable exposure
@@ -28,7 +30,9 @@ resolve_conflicts(const ParameterMap &parameters_default, const ParameterMap &pa
   }
 
   // overrides: prefer provided exposure
-  if (parameters_init.count("AeEnable") && parameters_init.at("AeEnable").get<bool>() &&
+  if (parameters_init.count("AeEnable") &&
+      (parameters_init.at("AeEnable").get_type() != rclcpp::ParameterType::PARAMETER_NOT_SET) &&
+      parameters_init.at("AeEnable").get<bool>() &&
       parameters_init.count("ExposureTime"))
   {
     // disable auto exposure
@@ -56,7 +60,9 @@ check_conflicts(const std::vector<rclcpp::Parameter> &parameters_new,
 
   // is auto exposure going to be enabled?
   const bool ae_enabled =
-    parameter_map.count("AeEnable") && parameter_map.at("AeEnable").get<bool>();
+    parameter_map.count("AeEnable") &&
+    (parameter_map.at("AeEnable").get_type() != rclcpp::ParameterType::PARAMETER_NOT_SET) &&
+    parameter_map.at("AeEnable").get<bool>();
   // are new parameters setting the exposure manually?
   const bool exposure_updated =
     std::find_if(parameters_new.begin(), parameters_new.end(), [](const rclcpp::Parameter &param) {

